Libraries/LibTLS/ClientHandshake.cpp

```diff
@@ -0,0 +1,613 @@
+    }
+    auto version = (Version)convert_between_host_and_network(*(const u16*)buffer.offset_pointer(res));
+
+    res += 2;
+    if (!supports_version(version))
+        return (i8)Error::NotSafe;
+
+    memcpy(m_context.remote_random, buffer.offset_pointer(res), sizeof(m_context.remote_random));
+    res += sizeof(m_context.remote_random);
+
+    u8 session_length = buffer[res++];
+    if (buffer.size() - res < session_length) {
+        dbg() << "not enough data for session id";
+        return (i8)Error::NeedMoreData;
+    }
+
+    if (session_length && session_length <= 32) {
+        memcpy(m_context.session_id, buffer.offset_pointer(res), session_length);
+        m_context.session_id_size = session_length;
+        dbg() << "Remote session ID:";
+        print_buffer(ByteBuffer::wrap(m_context.session_id, session_length));
+    } else {
+        m_context.session_id_size = 0;
+    }
+    res += session_length;
+
+    if (buffer.size() - res < 2) {
+        dbg() << "not enough data for cipher suite listing";
+        return (i8)Error::NeedMoreData;
+    }
+    auto cipher = (CipherSuite)convert_between_host_and_network(*(const u16*)buffer.offset_pointer(res));
+    res += 2;
+    if (!supports_cipher(cipher)) {
+        m_context.cipher = CipherSuite::Invalid;
+        dbg() << "No supported cipher could be agreed upon";
+        return (i8)Error::NoCommonCipher;
+    }
+    m_context.cipher = cipher;
+    dbg() << "Cipher: " << (u16)cipher;
+
+    // The handshake hash function is _always_ SHA256
+    m_context.handshake_hash.initialize(Crypto::Hash::HashKind::SHA256);
+
+    if (buffer.size() - res < 1) {
+        dbg() << "not enough data for compression spec";
+        return (i8)Error::NeedMoreData;
+    }
+    u8 compression = buffer[res++];
+    if (compression != 0) {
+        dbg() << "Server told us to compress, we will not!";
+        return (i8)Error::CompressionNotSupported;
+    }
+
+    if (res > 0) {
+        if (m_context.connection_status != ConnectionStatus::Renegotiating)
+            m_context.connection_status = ConnectionStatus::Negotiating;
+        if (m_context.is_server) {
+            dbg() << "unsupported: server mode";
+            write_packets = WritePacketStage::ServerHandshake;
+        }
+    }
+
+    if (res > 2) {
+        res += 2;
+    }
+
+    while ((ssize_t)buffer.size() - res >= 4) {
+        u16 extension_type = convert_between_host_and_network(*(const u16*)buffer.offset_pointer(res));
+        res += 2;
+        u16 extension_length = convert_between_host_and_network(*(const u16*)buffer.offset_pointer(res));
+        res += 2;
+
+        dbg() << "extension " << extension_type << " with length " << extension_length;
+        if (extension_length) {
+            if (buffer.size() - res < extension_length) {
+                dbg() << "not enough data for extension";
+                return (i8)Error::NeedMoreData;
+            }
+
+            // SNI
+            if (extension_type == 0x00) {
```

Comment:
I'm guessing these are some of the the TLS hello extensions that we support (from [the TLS extensiontype values website](http://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml), I guess this one would be `server_name` and the ones below would be `application_layer_protocol_negotiation` and `signature_algorithms`). Let's put that in an enum for clarity ?

Reply:
enums for days :^)